core/shared/src/main/scala/interface.scala

```diff
@@ -140,12 +169,25 @@ trait Param[Typeclass[_], Type] extends Serializable {
   *  @tparam Typeclass  type constructor for the typeclass being derived
   *  @tparam Type       generic type of this parameter */
 abstract class CaseClass[Typeclass[_], Type] (
-  val typeName: TypeName,
-  val isObject: Boolean,
-  val isValueClass: Boolean,
+  override val typeName: TypeName,
+  override val isObject: Boolean,
+  override val isValueClass: Boolean,
   parametersArray: Array[Param[Typeclass, Type]],
   annotationsArray: Array[Any]
-) extends Serializable {
+) extends ReadOnlyCaseClass[Typeclass, Type](
+  typeName,
+  isObject,
+  isValueClass,
+  // Safe to cast as we're never mutating the array
+  parametersArray.asInstanceOf[Array[ReadOnlyParam[Typeclass, Type]]],
```

Comment:
Hmm this made me realise that we are converting to a `Seq` on every method call and I think this is **not** efficient on Scala 2.13 because it does copy on write. I think we should change that (maybe in another PR). @propensive WDYT?

Reply:
Hmm. I actually have no idea what changed in Scala 2.13's treatment of arrays... What would the change be?

Reply:
Right now if you try to compile it warns:
```scala
[warn] ../magnolia/core/shared/src/main/scala/interface.scala:124:37: method copyArrayToImmutableIndexedSeq in class LowPriorityImplicits2 is deprecated (since 2.13.0): Implicit conversions from Array to immutable.IndexedSeq are implemented by copying; Use the more efficient non-copying ArraySeq.unsafeWrapArray or an explicit toIndexedSeq call
[warn]   final def annotations: Seq[Any] = annotationsArray
```